src/QueryFactory/FindObjectsFromRawSqlQueryFactory.php

```diff
@@ -130,12 +132,46 @@ private function processParsedUnionQuery(array $parsedSql, ?string $sqlCount): a
 
         $generator = new PHPSQLCreator();
 
-        $processedSql = $generator->create($query);
+        // Replaced the default generator by our own to add parenthesis around each SELECT
+        $processedSql = $this->buildUnion($query);
         $processedSqlCount = $generator->create($countQuery);
 
+        // Let's add the ORDER BY if any
+        if (isset($parsedSql['0']['ORDER'])) {
+            $orderByBuilder = new OrderByBuilder();
+            $processedSql .= " " . $orderByBuilder->build($parsedSql['0']['ORDER']);
+        }
+
         return [$processedSql, $sqlCount ?? $processedSqlCount, $columnDescriptors];
     }
 
+    /**
+     * @param mixed[] $parsed
+     */
+    private function buildUnion(array $parsed): string
+    {
+        $sql = '';
+        $select_builder = new SelectStatementBuilder();
+        $first = true;
+        foreach ($parsed['UNION'] as $clause) {
+            if (!$first) {
+                $sql .= ") UNION (";
+            } else {
+                $sql .= "(";
+                $first = false;
+            }
+
+            $sql .= $select_builder->build($clause);
+        }
+
+        // Close the parenthesis if needed
+        if (!$first) {
+            $sql .= ")";
+        }
```

Comment:
Can this be updated to a `array_map` and `implode` call ?

Reply:
done